### PR TITLE
Reduce data copies in Client::ReadObject.

### LINKED ISSUE #2748: Reduce data copying in ObjectReadStream.
There is excessive data copying in ObjectReadStream:

- We should take advantage of the [std::basic_streambuf<char>::xsgetn](https://en.cppreference.com/w/cpp/io/basic_streambuf/sgetn) function.
- Most of the time we should have the libcurl write callbacks write directly into the user buffer provided by `xsgetn`.
- When that is not possible, we should have the callback write directly to a the get area in the `ObjectReadStreambuf`.

### PR DESCRIPTION
**Pull request by coryan**
When the application uses `std::basic_stream<char>::read()` we can copy
the data received in the download straight into their buffer. And when
the application calls something like `getc()` we can still copy directly
into the internal buffer in the `ObjectReadStreambuf`.

I had to change the internals of `CurlDownloadRequest` to support this:
libcurl requires write callbacks (this is the callback for downloads,
strange naming, but whatevs) process the full buffer they are given.
When the application gives us a buffer in `.read()` that does not line
up perfectly with the write callback buffer we need to save the data
somewhere. I added a `spill` buffer in `CurlDownloadRequest` to keep
this data. These are typically small and we can use a fixed-size array
to support them.

This fixes #2748.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2759)

### CODE REVIEW

**Comment by coryan** [comment]
There is about a 4% improvement, that is good enough for me, though I was hoping for a lot more.

image

File: google/cloud/storage/internal/object_streambuf.cc [hunk 1]
```diff
@@ -108,6 +115,54 @@ ObjectReadStreambuf::int_type ObjectReadStreambuf::underflow() {
...
+    return offset;
+  }
+
+  StatusOr<ReadSourceResult> read_result =
+      source_->Read(s + offset, static_cast<std::size_t>(count - offset));
+  GCP_LOG(INFO) << __func__ << "() - count=" << count
+                << ", in_avail=" << in_avail() << ", offset=" << offset
+                << ", status=" << read_result.status();
+  // If there was an error set the internal state, but we still return the
+  // number of bytes
```
Comment (devbww) [inline google/cloud/storage/internal/object_streambuf.cc]: Nit: s/$/./

File: google/cloud/storage/internal/object_read_source.h [hunk 2]
```diff
@@ -25,6 +25,27 @@ namespace cloud {
...
+ *   headers may have been received too.
+ * - All (or part, or none) of the data requested was read, and the connection
+ *   is closed with some success HTTP response code, some headers may have been
+ *   received too.
+ * - All (or part, or none) of the data requested was read, and the connection
+ *   was closed with some error HTTP response code. Some error payload, and some
+ *   headers are part of this response.
+ * - There was an error trying to read the data: we wrap this object in a
+ *   StatusOr for this case.
+ * - Parts When reading data the result
```
Comment (devbww) [inline google/cloud/storage/internal/object_read_source.h]: Truncated description?

File: google/cloud/storage/internal/curl_download_request.cc [hunk 3]
```diff
@@ -28,76 +28,141 @@ namespace storage {
...
+      return std::move(running_handles).status();
+    }
+    // Only wait if there are CURL handles with pending work *and* the
+    // predicate is not satisfied. Note that if the predicate is ill-defined
+    // it might continue to be unsatisfied even though the handles have
+    // completed their work.
+    if (*running_handles == 0 || predicate()) {
+      return Status();
+    }
+    auto status = WaitForHandles(repeats);
```
Comment (devbww) [inline google/cloud/storage/internal/curl_download_request.cc]: Nit: It is slightly confusing that the argument is in/out, but that's just an opinion.

File: google/cloud/storage/internal/curl_download_request.cc [hunk 4]
```diff
@@ -28,76 +28,141 @@ namespace storage {
...
+    auto running_handles = PerformWork();
+    if (!running_handles.ok()) {
+      return std::move(running_handles).status();
+    }
+    // Only wait if there are CURL handles with pending work *and* the
+    // predicate is not satisfied. Note that if the predicate is ill-defined
+    // it might continue to be unsatisfied even though the handles have
+    // completed their work.
+    if (*running_handles == 0 || predicate()) {
+      return Status();
```
Comment (devbww) [inline google/cloud/storage/internal/curl_download_request.cc]: Nit: I might spell this as `break`.

Comment (coryan) [reply google/cloud/storage/internal/curl_download_request.cc]: Done.

File: google/cloud/storage/internal/curl_download_request.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_download_request.cc]: I somewhat agree, but it is a private function of a class in the internal namespace, I going to leave as-is.

File: google/cloud/storage/internal/object_read_source.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/object_read_source.h]: Thanks, I rewrote the whole section.

File: google/cloud/storage/internal/object_streambuf.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/object_streambuf.cc]: Fixed.

File: google/cloud/storage/internal/object_streambuf.cc [hunk 5]
```diff
@@ -108,6 +115,54 @@ ObjectReadStreambuf::int_type ObjectReadStreambuf::underflow() {
...
+  }
+
+  // Maybe the internal get area is enough to satisfy this request, no need to
+  // read more in that case:
+  auto from_internal = (std::min)(count, in_avail());
+  std::memcpy(s, gptr(), static_cast<std::size_t>(from_internal));
+  gbump(static_cast<int>(from_internal));
+  offset += from_internal;
+  if (offset >= count) {
+    GCP_LOG(INFO) << __func__ << "() - count=" << count
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/object_streambuf.cc]: Fixed.

File: google/cloud/storage/tests/object_media_integration_test.cc [hunk 6]
```diff
@@ -769,6 +769,69 @@ TEST_F(ObjectMediaIntegrationTest, ReadFromOffsetXml) {
...
+  auto const maximum_chunk_size = 256 * 1024L;
+  auto const minimum_chunk_size = 16;
+  std::vector<char> buffer(maximum_chunk_size);
+  std::uniform_int_distribution<int> chunk_size_generator(0,
+                                                          maximum_chunk_size);
+  do {
+    auto size = chunk_size_generator(generator_);
+    if (size < minimum_chunk_size) {
+      std::string line;
+      std::getline(stream, line);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/tests/object_media_integration_test.cc]: Fixed (I think, this is subtle).

File: google/cloud/storage/tests/object_media_integration_test.cc [hunk 7]
```diff
@@ -769,6 +769,69 @@ TEST_F(ObjectMediaIntegrationTest, ReadFromOffsetXml) {
...
+  }
+
+  StatusOr<ObjectMetadata> source_meta = client->InsertObject(
+      bucket_name, object_name, large_text, IfGenerationMatch(0));
+  ASSERT_STATUS_OK(source_meta);
+
+  EXPECT_EQ(object_name, source_meta->name());
+  EXPECT_EQ(bucket_name, source_meta->bucket());
+
+  // Create a iostream to read the object back.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/tests/object_media_integration_test.cc]: Fixed.

**Comment by coryan** [comment]
PTAL.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**